File: operations/database/classes/Matrix.py
import sys, string
import Config
import datetime
import re
import json

from classes import Lookups

class Matrix( ) :

	"""Tools for Handling the Creation of the Matrix Table"""
	
	def __init__( self, db, cursor ) :
		self.db = db
		self.cursor = cursor
		self.lookups = Lookups.Lookups( db, cursor )
		
		# Quick Lookup Hashes
		self.interactionTypesHASH = self.lookups.buildInteractionTypeHash( )
		self.historyOperationsHASH = self.lookups.buildHistoryOperationsHash( )
		self.participantTypeHASH = self.lookups.buildParticipantTypesHash( )
		self.participantRoleHASH = self.lookups.buildParticipantRoleHash( )
		self.annotationHASH = self.lookups.buildAnnotationHash( )
		self.participantTypeMappingHASH = self.lookups.buildParticipantTypeMappingHash( )
		self.attributeTypeHASH = self.lookups.buildAttributeTypeHASH( )
		self.ontologyTermHASH = self.lookups.buildAttributeOntologyTermHASH( )
		
		# Attribute Value Lookups
		self.attributeValues = { }
		
	def buildMatrix( self ) :
		
		"""Build out the Matrix Table One Interaction at a TIme"""
		
		self.cursor.execute( "SELECT interaction_id, dataset_id, interaction_type_id, interaction_state FROM " + Config.DB_IMS + ".interactions ORDER BY interaction_id ASC" )
		
		intCount = 0
		for row in self.cursor.fetchall( ) :
		
			intCount += 1
		
			typeName = self.interactionTypesHASH[str(row['interaction_type_id'])]
			
			interaction = [str(row['interaction_id'])]
			interaction.append( self.buildInteractionDetails( str(row['interaction_id']) ) )
			interaction.extend( [str(row['dataset_id']), str(row['interaction_type_id']), typeName, row['interaction_state']] )
			
			historyInfo = self.fetchLatestHistory( row['interaction_id'] )
			historyOperation = self.historyOperationsHASH[str(historyInfo['history_operation_id'])]
			interaction.extend( [historyInfo['modification_type'], str(historyInfo['history_operation_id']), historyOperation, historyInfo['history_comment']] )
			
			self.cursor.execute( "INSERT INTO " + Config.DB_IMS + ".matrix VALUES( %s, %s, %s, %s, %s, %s, %s, %s, %s, %s )", interaction )
			
			if (intCount % 10000) == 0 :
				self.db.commit( )
			
		self.db.commit( )
		
	def buildInteractionDetails( self, interactionID ) :
	
		"""Build a JSON array representing the details of an interaction for quick display purposes"""
		
		interactionDetails = { "interaction_id" : interactionID }
		interactionDetails["participants"] = self.buildParticipantDetails( interactionID )
		interactionDetails["attributes"] = self.buildAttributeDetails( interactionID )
		
		return json.dumps( interactionDetails, ensure_ascii=False ).encode( 'utf8' )
		
	def buildParticipantDetails( self, interactionID ) :
	
		"""Build a list of participant details for an interaction""" 
		
		participants = []
		self.cursor.execute( "SELECT participant_id, participant_role_id FROM " + Config.DB_IMS + ".interaction_participants WHERE interaction_participant_status='active' AND interaction_id=%s", [interactionID] )
		
		for row in self.cursor.fetchall( ) :
			participant = { "participant_id" : str(row['participant_id']) }
			participant["participant_role_id"] = str(row['participant_role_id'])
			participant["participant_role_name"] = self.participantRoleHASH[participant["participant_role_id"]]
			participant["participant_type_id"] = self.participantTypeMappingHASH[participant["participant_id"]]
			participant["participant_type_name"] = self.participantTypeHASH[participant["participant_type_id"]]
			participant["participant_annotation"] = self.annotationHASH[participant["participant_id"]]
			participants.append( participant )
			
		return participants
		
	def buildAttributeDetails( self, interactionID ) :
	
		"""Build a set of attributes for an interaction"""
		
		attributes = { }
		self.cursor.execute( "SELECT interaction_attribute_id, attribute_id, user_id FROM " + Config.DB_IMS + ".interaction_attributes WHERE interaction_id=%s AND interaction_attribute_parent='0' AND interaction_attribute_status='active'", [interactionID] )
		
		for row in self.cursor.fetchall( ) :
		
			attribute = self.buildAttribute( str(row['attribute_id']), str(row['user_id']), str(row['interaction_attribute_id']) )
			
			# NESTED STRUCTURE OF CATEGORIES -> TYPES -> ATTRIBUTES -> ATTRIBUTE_CHILDREN
			
			if attribute["attribute_type_category_name"].lower( ) not in attributes :
				attributes[attribute["attribute_type_category_name"].lower( )] = { }
				
			if attribute["attribute_type_name"].lower( ) not in attributes[attribute["attribute_type_category_name"].lower( )] :
				attributes[attribute["attribute_type_category_name"].lower( )][attribute["attribute_type_name"].lower( )] = []
			
			attributes[attribute["attribute_type_category_name"].lower( )][attribute["attribute_type_name"].lower( )].append( attribute )
			
		return attributes
		
	def buildAttribute( self, attributeID, userID, interactionAttributeID ) :
	
		"""Build out the details on a single attribute and its children"""
		
		attribute = { "attribute_id" : attributeID, "user_id" : userID }
		
		attributeDetails = self.fetchAttributeValue( str(attributeID) )
		
		attributeInfo = self.attributeTypeHASH[str(attributeDetails["attribute_type_id"])]
		attribute["attribute_type_id"] = attributeInfo["attribute_type_id"]
		attribute["attribute_type_name"] = attributeInfo["attribute_type_name"]
		attribute["attribute_type_shortcode"] = attributeInfo["attribute_type_shortcode"]
		attribute["attribute_type_category_id"] = attributeInfo["attribute_type_category_id"]
		attribute["attribute_type_category_name"] = attributeInfo["attribute_type_category_name"]
		
		# ATTRIBUTE VALUE
		attribute["attribute_value"] = attributeDetails["attribute_value"]
		
		# ONTOLOGY TERM ATTRIBUTES
		attribute["attribute_annotation"] = { }
		if attribute["attribute_type_category_id"] == 1 :
			attribute["attribute_annotation"] = self.ontologyTermHASH[attributeDetails["attribute_value"]]
		
		# CHILDREN
		self.cursor.execute( "SELECT interaction_attribute_id, attribute_id, user_id FROM " + Config.DB_IMS + ".interaction_attributes WHERE interaction_attribute_parent=%s AND interaction_attribute_status='active'", [interactionAttributeID] )
		
		attribute["attribute_children"] = []
		for row in self.cursor.fetchall( ) :
			attribute["attribute_children"].append( self.buildAttribute( str(row['attribute_id']), str(row['user_id']), str(row['interaction_attribute_id']) ))
			
		return attribute
			
	def fetchAttributeValue( self, attributeID ) :
	
		"""Fetch attribute value from HASH or look up in database"""
	
		if str(attributeID) not in self.attributeValues :
			self.cursor.execute( "SELECT attribute_value, attribute_type_id FROM " + Config.DB_IMS + ".attributes WHERE attribute_id=%s LIMIT 1", [attributeID] )
			row = self.cursor.fetchone( )
			self.attributeValues[str(attributeID)] = row
		
		return self.attributeValues[str(attributeID)]
		
	def fetchLatestHistory( self, interactionID ) :
	
		"""Grab latest history details from the history table"""
		
		self.cursor.execute( "SELECT modification_type, history_comment, history_operation_id FROM " + Config.DB_IMS + ".history WHERE interaction_id=%s ORDER BY history_addeddate DESC LIMIT 1", [interactionID] )
		
		row = self.cursor.fetchone( )
		return row

File: operations/elastic/classes/Matrix.php
<?php

/**
 * Datasets
 * This model class is for handling data processing for new
 * and existing datasets stored in the database.
 */

use \PDO;

require_once __DIR__ . '/../../../site/app/classes/models/Lookups.php';
 
class Matrix {

	private $db;
	private $lookups;
	
	private $interactionTypesHash;
	private $userNameHash;
	private $participantTypeHash;
	private $participantRoleHASH;
	private $annotationHASH;
	private $participantTypeMappingHASH;
	private $attributeTypeHASH;
	private $ontologyTermHASH;
	
	private $hosts;
	private $clientBuilder;
	private $client;
	
	private $BULK_BATCH_SIZE = 500;

	/**
	 * Establish a database connection and also build several quick
	 * lookup hashes so we can speed up the process.
	 */
	
	public function __construct( ) {
		$this->db = new PDO( DB_CONNECT, DB_USER, DB_PASS );
		$this->db->setAttribute( PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION );
		
		$this->lookups = new IMS\app\classes\models\Lookups( );
		
		$this->hosts = array( ES_HOST . ":" . ES_PORT );
		$this->clientBuilder = Elasticsearch\ClientBuilder::create( );
		$this->clientBuilder->setHosts( $this->hosts );
		$this->client = $this->clientBuilder->build( );
	}
	
	/**
	 * Build quick lookups for batch processing
	 */
	 
	private function buildQuickLookups( ) {
		$this->interactionTypesHash = $this->lookups->buildInteractionTypeHash( );
		$this->userNameHash = $this->lookups->buildUserNameHash( );
		$this->participantTypeHash = $this->lookups->buildParticipantTypesHash( );
		$this->participantRoleHash = $this->lookups->buildParticipantRoleHash( );
		$this->annotationHash = $this->lookups->buildAnnotationHash( );
		$this->participantTypeMappingHash = $this->lookups->buildParticipantTypeMappingHash( );
		$this->attributeTypeHash = $this->lookups->buildAttributeTypeHASH( );
		$this->ontologyTermHash = $this->lookups->buildAttributeOntologyTermHASH( );
	}
	
	/**
	 * Build the matrix by fetching interactions by dataset id
	 */
	 
	public function buildMatrixByDataset( $datasetID ) {
		
		$this->buildQuickLookups( );
		
		$stmt = $this->db->prepare( "SELECT * FROM " . DB_IMS . ".interactions WHERE dataset_id=?" );
		$stmt->execute( array( $datasetID ) );
		
		// Build Param Format
		$params = array( 
			"body" => array( )
		);
		
		$documentCount = 0;
		while( $row = $stmt->fetch( PDO::FETCH_OBJ ) ) {
			
			$documentCount++;
			
			$params["body"][] = array( 
				"index" => array( 
					"_index" => ES_INDEX,
					"_type" => ES_TYPE,
					"_id" => $row->interaction_id
				)
			);
			
			// Build Document
			$document = $this->buildMatrixDocument( $row );
			$params["body"][] = $document;

			if( ($documentCount % $this->BULK_BATCH_SIZE) == 0 ) {
				// Insert it into Elastic Search
				$responses = $this->client->bulk( $params );
				$params["body"] = array( );
				print_r( $responses );
			}
			
		}
		
		if( sizeof( $params["body"] ) > 0 ) {
			$responses = $this->client->bulk( $params );
			$params["body"] = array( );
			print_r( $responses );
		}
	}
	
	/**
	 *  Build the matrix by fetching interactions by interaction_id
	 */
	 
	public function buildMatrixByInteraction( $interactionID ) {
		
		$this->buildQuickLookups( );
		
		$stmt = $this->db->prepare( "SELECT * FROM " . DB_IMS . ".interactions WHERE interaction_id=?" );
		$stmt->execute( array( $interactionID ) );
		
		while( $row = $stmt->fetch( PDO::FETCH_OBJ ) ) {
			// Build Document
			$document = $this->buildMatrixDocument( $row );

			// Build Param Format
			$params = array( 
				"index" => ES_INDEX,
				"type" => ES_TYPE,
				"id" => $row->interaction_id,
				"body" => $document
			);
			
			// Insert it into Elastic Search
			$response = $this->client->index( $params );
				
		}
		
	}
	
	/**
	 * Build a matrix document using the structure required by our search system
	 */
	 
	private function buildMatrixDocument( $interaction ) {
		
		$document = array( );
		$document['interaction_id'] = $interaction->interaction_id;
		$document['dataset_id'] = $interaction->dataset_id;
		$document['interaction_type_id'] = $interaction->interaction_type_id;
		$document['interaction_type_name'] = $this->interactionTypesHash[$interaction->interaction_type_id];
		$document['interaction_state'] = $interaction->interaction_state;
		
		$document += $this->fetchInteractionHistoryDetails( $interaction->interaction_id );
		$document['attributes'] = $this->fetchInteractionAttributes( $interaction->interaction_id );
		$document['participants'] = $this->fetchInteractionParticipants( $interaction->interaction_id );
		
		return $document;
		
	}
	
	/**
	 * Return an array containing interaction history details for mapping to a document
	 */
	 
	private function fetchInteractionHistoryDetails( $interactionID ) {
		
		$stmt = $this->db->prepare( "SELECT modification_type, user_id, history_addeddate FROM " . DB_IMS . ".history WHERE interaction_id=? AND modification_type IN ('ACTIVATED','DISABLED') ORDER BY history_addeddate DESC LIMIT 1" );
		$stmt->execute( array( $interactionID ) );
		
		$historyDetails = array( ); 
		if( $row = $stmt->fetch( PDO::FETCH_OBJ ) ) {
			$historyDetails['history_status'] = $row->modification_type;
			$historyDetails['history_user_id'] = $row->user_id;
			$historyDetails['history_user_name'] = $this->userNameHash[$row->user_id];
			$historyDetails['history_date'] = date( 'Y/m/d H:i:s', strtotime( $row->history_addeddate ));
		}
		
		return $historyDetails;
		
	}
	
	/**
	 * Return an array containing interaction attributes
	 */
	
	private function fetchInteractionAttributes( $interactionID ) {
	
		$stmt = $this->db->prepare( "SELECT interaction_attribute_id, attribute_id, interaction_attribute_parent, user_id, interaction_attribute_addeddate FROM " . DB_IMS . ".interaction_attributes WHERE interaction_id=? AND interaction_attribute_status='active' ORDER BY interaction_attribute_addeddate ASC" );
		$stmt->execute( array( $interactionID ) );
		
		$attDetails = array( );		
		while( $row = $stmt->fetch( PDO::FETCH_OBJ ) ) {
			$attDetail = array( );
			$attDetail['interaction_attribute_id'] = $row->interaction_attribute_id;
			$attDetail['attribute_id'] = $row->attribute_id;
			$attDetail['attribute_parent_id'] = $row->interaction_attribute_parent;
			$attDetail['attribute_user_id'] = $row->user_id;
			$attDetail['attribute_user_name'] = $this->userNameHash[$row->user_id];
			$attDetail['attribute_addeddate'] = date( 'Y/m/d H:i:s', strtotime( $row->interaction_attribute_addeddate ));
			
			$attDetail += $this->fetchAttribute( $row->attribute_id );
			$attDetails[] = $attDetail;
		}
		
		return $attDetails;
	
	}
	
	/**
	 * Return an array of attribute details
	 */
	
	private function fetchAttribute( $attributeID ) {
	
		$stmt = $this->db->prepare( "SELECT attribute_value, attribute_type_id FROM " . DB_IMS . ".attributes WHERE attribute_id=? LIMIT 1" );
		$stmt->execute( array( $attributeID ) );
		
		$attribute = array( ); 
		if( $row = $stmt->fetch( PDO::FETCH_OBJ ) ) {
			$attribute['attribute_value'] = $row->attribute_value;
			$attribute['attribute_type_id'] = $row->attribute_type_id;
			
			$attributeTypeInfo = $this->attributeTypeHash[$row->attribute_type_id];
			$attribute['attribute_type_name'] = $attributeTypeInfo->attribute_type_name;
			$attribute['attribute_type_shortcode'] = $attributeTypeInfo->attribute_type_shortcode;
			$attribute['attribute_type_category_id'] = $attributeTypeInfo->attribute_type_category_id;
			$attribute['attribute_type_category_name'] = $attributeTypeInfo->attribute_type_category_name;
			$attribute['ontology_term_id'] = '0';
			$attribute['ontology_term_official_id'] = '0';
			
			// FOR ONTOLOGY TERMS ONLY
			if( $attribute['attribute_type_category_id'] == "1" ) {
				$ontologyInfo = $this->ontologyTermHash[$row->attribute_value];
				$attribute['attribute_value'] = $ontologyInfo->ontology_term_name;
				$attribute['ontology_term_id'] = $row->attribute_value;
				$attribute['ontology_term_official_id'] = $ontologyInfo->ontology_term_official_id;
			}
			
		}
		
		return $attribute;
	
	}
	
	/**
	 * Return an array containing interaction participants
	 */
	
	private function fetchInteractionParticipants( $interactionID ) {
	
		$stmt = $this->db->prepare( "SELECT interaction_participant_id, participant_id, participant_role_id FROM " . DB_IMS . ".interaction_participants WHERE interaction_id=? AND interaction_participant_status='active'" );
		$stmt->execute( array( $interactionID ) );
		
		$partDetails = array( );		
		while( $row = $stmt->fetch( PDO::FETCH_OBJ ) ) {
			$partDetail = array( );
			$partDetail['interaction_participant_id'] = $row->interaction_participant_id;
			$partDetail['participant_id'] = $row->participant_id;
			$partDetail['participant_role_id'] = $row->participant_role_id;
			$partDetail['participant_role_name'] = $this->participantRoleHash[$row->participant_role_id];
			$partDetail['participant_type_id'] = $this->participantTypeMappingHash[$row->participant_id];
			$partDetail['participant_type_name'] = $this->participantTypeHash[$partDetail['participant_type_id']];
			
			$partDetail += $this->fetchParticipantAnnotation( $row->participant_id );
			
			// NEED TO ADD PARTICIPANT ATTRIBUTES HERE AS WELL FOR ALLELES
			$partDetail['attributes'] = $this->fetchInteractionParticipantAttributes( $row->interaction_participant_id );
			
			$partDetails[] = $partDetail;
		}
		
		return $partDetails;
	
	}
	
	/**
	 * Return an array of annotation for a given participant
	 */
	
	private function fetchParticipantAnnotation( $participantID ) {
		return $this->annotationHash[$participantID];
	}
	
	/**
	 * Return an array containing interaction participant attributes
	 */
	
	private function fetchInteractionParticipantAttributes( $interactionParticipantID ) {
	
		$stmt = $this->db->prepare( "SELECT interaction_participant_attribute_id, attribute_id, interaction_participant_attribute_parent, user_id, interaction_participant_attribute_addeddate FROM " . DB_IMS . ".interaction_participant_attributes WHERE interaction_participant_id=? AND interaction_participant_attribute_status='active' ORDER BY interaction_participant_attribute_addeddate ASC" );
		$stmt->execute( array( $interactionParticipantID ) );
		
		$attDetails = array( );		
		while( $row = $stmt->fetch( PDO::FETCH_OBJ ) ) {
			$attDetail = array( );
			$attDetail['interaction_participant_attribute_id'] = $row->interaction_participant_attribute_id;
			$attDetail['attribute_id'] = $row->attribute_id;
			$attDetail['attribute_parent_id'] = $row->interaction_participant_attribute_parent;
			$attDetail['attribute_user_id'] = $row->user_id;
			$attDetail['attribute_user_name'] = $this->userNameHash[$row->user_id];
			$attDetail['attribute_addeddate'] = date( 'Y/m/d H:i:s', strtotime( $row->interaction_participant_attribute_addeddate ));
			
			$attDetail += $this->fetchAttribute( $row->attribute_id );
			$attDetails[] = $attDetail;
		}
		
		return $attDetails;
	
	}
	
	/**
	 * Initialize the publications index from scratch
	 */
	 
	public function initialize( ) {
		
		// DELETE EXISTING INDEX

		$params = array( 
			"index" => ES_INDEX
		);

		$response = $this->client->indices( )->delete( $params );
		print_r( $response );
		
		// CREATE NEW INDEX

		$params = array( 
			"index" => ES_INDEX,
			"body" => array( 
				"settings" => array( 
					"number_of_shards" => 1,
					"number_of_replicas" => 0,
					"index" => array(
						"analysis" => array(
							"analyzer" => array( 
								"keyword_analyzer" => array(
									"tokenizer" => "keyword",
									"filter" => "lowercase"
								)
							)
						)
					)
				)
			)
		);
		
		// ADD A MAPPING
		
		$response = $this->client->indices( )->create( $params );
		print_r( $response );
		
		$params = array( 
			"index" => ES_INDEX,
			"type" => ES_TYPE,
			"body" => array( 
				"interaction" => array( 
					"_all" => array( "enabled" => true ),
					"_source" => array( "enabled" => true ),
					"properties" => array( 
						"interaction_id" => array( "type" => "integer" ),
						"dataset_id" => array( "type" => "integer" ),
						"interaction_type_id" => array( "type" => "integer" ),
						"interaction_state" => array( "type" => "string", "analyzer" => "keyword_analyzer" ),
						"history_status" => array( "type" => "string", "analyzer" => "keyword_analyzer" ),
						"history_user_id" => array( "type" => "integer" ),
						"history_user_name" => array( "type" => "string" ),
						"history_date" => array( "type" => "date", "format" => "yyyy/MM/dd HH:mm:ss" ),
						"attributes" => array( "type" => "nested", "properties" => array(  
							"interaction_attribute_id" => array( "type" => "integer" ),
							"attribute_id" => array( "type" => "integer" ),
							"attribute_parent_id" => array( "type" => "integer" ),
							"attribute_user_id" => array( "type" => "integer" ),
							"attribute_user_name" => array( "type" => "string" ),
							"attribute_addeddate" => array( "type" => "date", "format" => "yyyy/MM/dd HH:mm:ss" ),
							"attribute_value" => array( "type" => "string", "analyzer" => "keyword_analyzer", "fields" => array( 
								"full" => array( "type" => "string" )
							)),
							"attribute_type_id" => array( "type" => "integer" ),
							"attribute_type_name" => array( "type" => "string", "analyzer" => "keyword_analyzer" ),
							"attribute_type_shortcode" => array( "type" => "string", "analyzer" => "keyword_analyzer" ),
							"attribute_type_category_id" => array( "type" => "integer" ),
							"attribute_type_category_name" => array( "type" => "string", "analyzer" => "keyword_analyzer" ),
							"ontology_term_id" => array( "type" => "integer" ),
							"ontology_term_official_id" => array( "type" => "string", "analyzer" => "keyword_analyzer" )
						)),
						"participants" => array( "type" => "nested", "properties" => array( 
							"interaction_participant_id" => array( "type" => "integer" ),
							"participant_id" => array( "type" => "integer" ),
							"participant_role_id" => array( "type" => "integer" ),
							"participant_role_name" => array( "type" => "string", "analyzer" => "keyword_analyzer" ),
							"participant_type_id" => array( "type" => "integer" ),
							"participant_type_name" => array( "type" => "string", "analyzer" => "keyword_analyzer" ),
							"primary_name" => array( "type" => "string", "analyzer" => "keyword_analyzer", "fields" => array( 
								"full" => array( "type" => "string" )
							)),
							"systematic_name" => array( "type" => "string", "analyzer" => "keyword_analyzer", "fields" => array( 
								"full" => array( "type" => "string" )
							)),
							"aliases" => array( "type" => "string", "analyzer" => "keyword_analyzer", "fields" => array( 
								"full" => array( "type" => "string" )
							)),
							"organism_id" => array( "type" => "integer" ),
							"organism_official_name" => array( "type" => "string", "analyzer" => "keyword_analyzer" ),
							"organism_abbreviation" => array( "type" => "string", "analyzer" => "keyword_analyzer" ),
							"organism_strain" => array( "type" => "string", "analyzer" => "keyword_analyzer" ),
							"attributes" => array( "type" => "nested", "properties" => array( 
								"interaction_attribute_id" => array( "type" => "integer" ),
								"attribute_id" => array( "type" => "integer" ),
								"attribute_parent_id" => array( "type" => "integer" ),
								"attribute_user_id" => array( "type" => "integer" ),
								"attribute_user_name" => array( "type" => "string", "analyzer" => "keyword_analyzer" ),
								"attribute_addeddate" => array( "type" => "date", "format" => "yyyy/MM/dd HH:mm:ss" ),
								"attribute_value" => array( "type" => "string", "analyzer" => "keyword_analyzer", "fields" => array( 
									"full" => array( "type" => "string" )
								)),
								"attribute_type_id" => array( "type" => "integer" ),
								"attribute_type_name" => array( "type" => "string", "analyzer" => "keyword_analyzer" ),
								"attribute_type_shortcode" => array( "type" => "string", "analyzer" => "keyword_analyzer" ),
								"attribute_type_category_id" => array( "type" => "integer" ),
								"attribute_type_category_name" => array( "type" => "string", "analyzer" => "keyword_analyzer" ),
								"ontology_term_id" => array( "type" => "integer" ),
								"ontology_term_official_id" => array( "type" => "string", "analyzer" => "keyword_analyzer" )
							))
						))
					)
				)
			)
		);
		
		$response = $this->client->indices( )->putMapping( $params );
		print_r( $response );
		
	}
	
	/**
	 * Submit a search to elastic search and return the response
	 */
	 
	public function search( $params ) {
		return $this->client->search( $params );
	}
	
}

?>